sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -94,56 +95,78 @@ public async ValueTask<string> GetHeaderValueAsync(HttpMessage message, bool asy
+                            headerValueTcs.SetResult(info);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            headerValueTcs.SetCanceled();
+                            throw;
+                        }
+                        catch (Exception exception)
+                        {
+                            headerValueTcs.SetException(exception);
+                            throw;
+                        }
                     }
 
+                    var headerValueTask = headerValueTcs.Task;
                     try
                     {
-                        HeaderValueInfo info = await GetHeaderValueFromCredentialAsync(message, async, message.CancellationToken).ConfigureAwait(false);
-                        headerValueTcs.SetResult(info);
-                    }
-                    catch (OperationCanceledException)
-                    {
-                        headerValueTcs.SetCanceled();
-                        throw;
-                    }
-                    catch (Exception exception)
-                    {
-                        headerValueTcs.SetException(exception);
-                        throw;
-                    }
-                }
+                        if (!headerValueTask.IsCompleted)
+                        {
+                            if (async)
+                            {
+                                await headerValueTask.AwaitWithCancellation(message.CancellationToken);
+                            }
+                            else
+                            {
+                                try
+                                {
+                                    headerValueTask.Wait(message.CancellationToken);
+                                }
+                                catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+                            }
+                        }
 
-                var headerValueTask = headerValueTcs.Task;
-                if (!headerValueTask.IsCompleted)
-                {
-                    if (async)
-                    {
-                        await headerValueTask.AwaitWithCancellation(message.CancellationToken);
+#pragma warning disable AZC0111 // DO NOT use EnsureCompleted in possibly asynchronous scope.
+                        return headerValueTcs.Task.EnsureCompleted().HeaderValue;
+#pragma warning restore AZC0111 // DO NOT use EnsureCompleted in possibly asynchronous scope.
                     }
-                    else
+
+                    catch (TaskCanceledException)
                     {
-                        try
+                        // We were waiting on a previous headerValueTcs operation which was canceled.
+
+                        // If we have already retired once, throw.
+                        if (hasRetriedDueToCancellation)
```

Comment:
Shouldn't we keep retrying while the current request is not canceled?

Reply:
We end up here if the current request was cancelled or the previous TCS that we were waiting on to complete was canceled. Neither scenario should be retried as part of this policy - that would be for the Retry policy to handle, if one is configured.

Reply:
I should clarify - We only retry in the scenario where the previous TCS we were waiting on to complete was canceled, but our current request has not been canceled. This is to honor the intent of the current request's cancellation status rather than letting a previous request's cancellation dictate that the current request should also be canceled.

Reply:
True, but while we are looping back to `(headerValueTcs, backgroundUpdateTcs, getTokenFromCredential) = GetTaskCompletionSources();` another thread may have started acquiring a token. There is no guarantee that on the next iteration it would be the current thread requesting the token.

Reply:
Ah, I understand now. Yes, I think you are correct.

Reply:
However, we do need some mechanism to avoid retying forever, in the event that the request has no cancellation token set.

Reply:
I wonder if at some point we should just force the current thread to independently get a token.

Reply:
> I wonder if at some point we should just force the current thread to independently get a token.

I'm not sure if it is worth adding the complexity considering this would have to be a very specific scenario to retry more than once due to the currently active tasks cancellation.

Reply:
>I'm not sure if it is worth

Yeah, same.